bdai_ros2_wrappers/bdai_ros2_wrappers/executors.py

```diff
@@ -0,0 +1,414 @@
+# Copyright (c) 2023 Boston Dynamics AI Institute Inc.  All rights reserved.
+import atexit
+import collections
+import concurrent.futures
+import dataclasses
+import logging
+import os
+import queue
+import threading
+import typing
+import weakref
+
+import rclpy.executors
+
+
+def classname(obj: typing.Any) -> str:
+    cls = obj.__class__
+    return f"{cls.__module__}.{cls.__qualname__}"
+
+
+class AutoScalingThreadPool(concurrent.futures.Executor):
+    """
+    A concurrent.futures.Executor subclass based on a thread pool.
+
+    Akin to the concurrent.futures.ThreadPoolExecutor class but with
+    autoscaling capabilities.
+    """
+
+    _lock: threading.Lock = threading.Lock()
+    _interpreter_shutdown: bool = False
+    _all_runqueues: weakref.WeakSet = weakref.WeakSet()
+    _all_workers: weakref.WeakSet = weakref.WeakSet()
+
+    @classmethod
+    def _on_interpreter_shutdown(cls) -> None:
+        # Gracefully shutdown daemonized threads
+        # upon interpreter shutdown.
+        with cls._lock:
+            cls._interpreter_shutdown = True
+            for q in cls._all_runqueues:
+                q.put(None)
+            cls._all_runqueues.clear()
+            for worker in cls._all_workers:
+                if worker.is_alive():
+                    worker.join()
+            cls._all_workers.clear()
+
+    @dataclasses.dataclass
+    class Work:
+        future: concurrent.futures.Future
+        fn: typing.Callable[..., typing.Any]
+        args: typing.Tuple[typing.Any, ...]
+        kwargs: typing.Dict[str, typing.Any]
+
+        def execute(self) -> None:
+            if not self.future.set_running_or_notify_cancel():
+                return
+            try:
+                self.future.set_result(self.fn(*self.args, **self.kwargs))
```

Comment:
Right, it would be an issue, but it cannot happen in this implementation. Cancellations are notified for work that is cancelled while in _waitqueues_, while work is still out of reach for workers that can execute it. I've extended the documentation to discuss these waitqueues.